quality-assurance/core.md

```diff
@@ -0,0 +1,193 @@
+* Components are split out logically where appropriate, make use of Laravel's powerful blade templating system if 
+appropriate (clientCard, navbar etc)
+
+**Service Providers**
+* Service Providers are utilised where appropriate to bind classes into the container (API interfaces etc)
+
+**File Storage**
+* Files are stored on S3 unless requested by the client to be stored locally or agreed beforehand
+* Ensure appropriate permissions on files in applicable (sensitive documents are private, avatars are public etc)
+
+**Queues**
+* Queues use Redis unless agreed beforehand
+* Ensure appropriate measures are in place for handling failed jobs
+
+**Broadcasting**
+* Broadcasting uses Laravel Echo unless agreed beforehand
+* Ensure appropriate authentication & security for each channel
+
+**Database**
+* All migrations use foreign keys for relations to ensure database integrity
+* SQL data types are appropriate
+* Monetary values should either be stored in pence as an INTEGER or as a DECIMAL - Never a FLOAT
+* Large tables that are queried often are indexed in order to keep database speed up
+
+## Processes
+### Automated Testing
+
+Automated testing should be implemented in a way to confirm functionality specified by the client.
+To this end, automated tests should be created to confirm the project user stories and related 
+business logic is satisfied and correct. 
+
+#### Unit Testing
+
+Our current policy is that unit tests should be used for:
+
+* Testing of project business logic / rules, that are able to be sufficient abstracted
+* Testing of packages, including our open source work
+
+#### Integration, Browser &amp; End-to-End Testing
+
+Integration, browser and end-to-end tests should be used for:
+
+* Testing of project functionality related to client specified user stories
+* Testing of critical path functionality, such as the register, login, and checkout process of an
+ecommerce site
+
+Since end-to-end tests are slower than browser tests and browser tests are slower than integration 
+tests, you should ideally aim to use the fastest possible test type, if appropriate.
+### Manual Testing
+#### 1 - Developer Testing
+During the development of the project, developers will create `feature/*` branches for all features going into the software.
+After the initial feature development - the developer should manually test their feature after writing passing 
+[automation tests](#automated-testing). This phase should also include a check that ensures the feature does not infringe
+any of the items listed in the ["What makes a good web application?"](#what-makes-a-good-web-application) section. 
+
+Once the feature has the thumbs up from the developer - that is when the pull request into the `master` branch should be
+made. 
+
+The developers should log any bugs found in this phase into JIRA to improve their ratio of bugs they find VS bugs that
+get through to the QA team/customer. 
+
+#### 2 - Sprint Testing (On the master branch)
```

Comment:
A possible problem  count occur here if QA is testing on the `master` branch, and development continues on with development of the next sprint in the mean time.

Reply:
@wheatleyjj What about testing on staging? We should be aiming to push a version to staging at the end of any sprint so it makes sense to do QA on that branch.

Reply:
I agree with @AlexCatch about testing on staging.

Reply:
I've worked on project previously with this kind of QA process in place. 

At the start of a sprint, a `sprint-1` branch would be created. After the sprint is over, QA would begin testing on `sprint-1`. At the same time a `sprint-2` branch would be created (from `sprint-1`) to allow developers to continue development immediately.

Once QA was happy with `sprint-1` it would be merged into `mater`, and the process would repeat.

Reply:
@ash123456789 @AlexCatch I think that testing on the customers environment would cause annoyances to the client if they're using it regularly. 

Ways around this are sprint branches as I and @DivineOmega suggested or having a development environment - but that comes with another layer of grief. I'm also thinking experienced QA testers will have good knowledge of automation tests so would be able to look into both the Blackbox/Whitebox sections of this document.

Thumbs up if you agree?

Reply:
@DivineOmega Agree with the sprint branches and/or development server idea. :+1: